buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy

```diff
@@ -35,91 +36,189 @@ import org.gradle.api.tasks.Exec
-        File pidFile = pidFile(baseDir)
+    /** Adds a dependency on the given distribution */
+    static void configureDistributionDependency(Project project, String distro) {
+        String elasticsearchVersion = ElasticsearchProperties.version
+        String packaging = distro == 'tar' ? 'tar.gz' : distro
+        project.configurations {
+            elasticsearchDistro
+        }
+        project.dependencies {
+            elasticsearchDistro "org.elasticsearch.distribution.${distro}:elasticsearch:${elasticsearchVersion}@${packaging}"
+        }
+    }
+
+    /**
+     * Adds dependent tasks to start an elasticsearch cluster before the given task is executed,
+     * and stop it after it has finished executing.
+     *
+     * The setup of the cluster involves the following:
+     * <ol>
+     *   <li>Cleanup the extraction directory</li>
+     *   <li>Extract a fresh copy of elasticsearch</li>
+     *   <li>Write an elasticsearch.yml config file</li>
+     *   <li>Copy plugins that will be installed to a temporary dir (which contains spaces)</li>
+     *   <li>Install plugins</li>
+     *   <li>Run additional setup commands</li>
+     *   <li>Start elasticsearch<li>
+     * </ol>
+     */
+    static void configureTasks(Project project, Task task, ClusterConfiguration config, File baseDir) {
         String clusterName = "${task.path.replace(':', '_').substring(1)}"
+        File pidFile = pidFile(baseDir)
         File home = homeDir(baseDir, config.distribution)
+        File cwd = new File(baseDir, "cwd")
+        File pluginsTmpDir = new File(baseDir, "plugins tmp")
+
+        // tasks are chained so their execution order is maintained
+        Task setup = project.tasks.create(name: "${task.name}#clean", type: Delete, dependsOn: task.dependsOn.collect()) {
+            delete home
+            doLast {
+                cwd.mkdirs()
+            }
+        }
+        setup = configureCheckPreviousTask("${task.name}#checkPrevious", project, setup, pidFile)
+        setup = configureStopTask("${task.name}#stopPrevious", project, setup, pidFile)
+        setup = configureExtractTask("${task.name}#extract", project, setup, baseDir, config.distribution)
+        setup = configureWriteConfigTask("${task.name}#configure", project, setup, home, config, clusterName, pidFile)
+        setup = configureCopyPluginsTask("${task.name}#copyPlugins", project, setup, pluginsTmpDir, config)
+
+        // install plugins
+        for (Map.Entry<String, FileCollection> plugin : config.plugins.entrySet()) {
+            String camelName = plugin.getKey().replaceAll(/-(\w)/) { _, c -> c.toUpperCase() }
+            String taskName = "${task.name}#install${camelName.capitalize()}"
+            // delay reading the file location until execution time
+            String file = "${ -> new File(pluginsTmpDir, plugin.getValue().singleFile.getName()).toURI().toURL().toString() }"
+            Object[] args = [new File(home, 'bin/plugin'), 'install', file]
```

Comment:
`${}` is groovy substitution in strings (essentially a closure that is executed when the string is rendered with a toString() call). In this particular case, I'm using it as a "lazy evaluated toString", so that the `singleFile` is not looked up until execution time (because it is filled in by the dependent task that builds the plugin).

I'm adding some comments to the couple places that use "nasty groovy", and I will do a follow up to make the code in general look less groovy like.